[Http-Client] "Failure when receiving data from the peer" / "Proxy CONNECT aborted" with v4.4.38+, OK with 4.4.37

### Symfony version(s) affected

4.4.38+

### Description

With http-client:4.4.38 and above (up to http-client:4.4.46), I don't manage to connect to an external API through our proxy:

```
$ php bin/console myapp:test-xxx
^ """
* About to connect() to proxy proxy.xxx.yyy.zzz port 8080 (#0)\n
*   Trying 1.2.3.4...\n
* TCP_NODELAY set\n
* Connected to proxy.xxx.yyy.zzz (1.2.3.4) port 8080 (#0)\n
* Establish HTTP proxy tunnel to api.provider.com:443\n
> CONNECT api.provider.com:443 HTTP/1.1\r\n
Host: api.provider.com:443\r\n
Proxy-Connection: Keep-Alive\r\n
Content-Type: application/json\r\n
Accept: application/json\r\n
Content-Length: 75\r\n
User-Agent: Symfony HttpClient/Curl\r\n
Accept-Encoding: gzip\r\n
\r\n
* Proxy CONNECT aborted\n
* Connection #0 to host proxy.xxx.yyy.zzz left intact\n
"""
14:10:48 ERROR     [app] XXX: post('') Failure when receiving data from the peer for "https://api.provider.com/". ["exception" => Symfony\Component\HttpClient\Exception\TransportException^ { …}]


 [ERROR] XXX : Failure when receiving data from the peer for
         "https://api.provider.com/".
```

### How to reproduce

Sorry, I didn't write any reproduction script, my use of CurlHttpClient seems to be completely standard and I manage to use other APIs within my application, the only thing that changes is HttpClient version.

### Possible Solution

If i downgrade to http-client:4.4.37, it does work:

$ php bin/console myapp:test-xxx

```
 [OK] XXX : OK
```

If i use directly curl_init(), curl_setopt() and curl_exec(), it does work, too.

If don't see what introduced this behaviour between the two releases looking at <https://github.com/symfony/http-client/compare/v4.4.37...v4.4.38>.

### Additional Context

NB: seems the same as #46499, I hope I can help resolving this regression.

--- a/src/Symfony/Component/HttpClient/CurlHttpClient.php
+++ b/src/Symfony/Component/HttpClient/CurlHttpClient.php
@@ -204,8 +204,14 @@ final class CurlHttpClient implements HttpClientInterface, LoggerAwareInterface,
         if (\extension_loaded('zlib') && !isset($options['normalized_headers']['accept-encoding'])) {
             $options['headers'][] = 'Accept-Encoding: gzip'; // Expose only one encoding, some servers mess up when more are provided
         }
+        $body = $options['body'];
 
-        foreach ($options['headers'] as $header) {
+        foreach ($options['headers'] as $i => $header) {
+            if (\is_string($body) && '' !== $body && 0 === stripos($header, 'Content-Length: ')) {
+                // Let curl handle Content-Length headers
+                unset($options['headers'][$i]);
+                continue;
+            }
             if (':' === $header[-2] && \strlen($header) - 2 === strpos($header, ': ')) {
                 // curl requires a special syntax to send empty headers
                 $curlopts[\CURLOPT_HTTPHEADER][] = substr_replace($header, ';', -2);
@@ -221,7 +227,7 @@ final class CurlHttpClient implements HttpClientInterface, LoggerAwareInterface,
             }
         }
 
-        if (!\is_string($body = $options['body'])) {
+        if (!\is_string($body)) {
             if (\is_resource($body)) {
                 $curlopts[\CURLOPT_INFILE] = $body;
             } else {
@@ -233,15 +239,16 @@ final class CurlHttpClient implements HttpClientInterface, LoggerAwareInterface,
             }
 
             if (isset($options['normalized_headers']['content-length'][0])) {
-                $curlopts[\CURLOPT_INFILESIZE] = substr($options['normalized_headers']['content-length'][0], \strlen('Content-Length: '));
-            } elseif (!isset($options['normalized_headers']['transfer-encoding'])) {
-                $curlopts[\CURLOPT_HTTPHEADER][] = 'Transfer-Encoding: chunked'; // Enable chunked request bodies
+                $curlopts[\CURLOPT_INFILESIZE] = (int) substr($options['normalized_headers']['content-length'][0], \strlen('Content-Length: '));
+            }
+            if (!isset($options['normalized_headers']['transfer-encoding'])) {
+                $curlopts[\CURLOPT_HTTPHEADER][] = 'Transfer-Encoding:'.(isset($curlopts[\CURLOPT_INFILESIZE]) ? '' : ' chunked');
             }
 
             if ('POST' !== $method) {
                 $curlopts[\CURLOPT_UPLOAD] = true;
 
-                if (!isset($options['normalized_headers']['content-type'])) {
+                if (!isset($options['normalized_headers']['content-type']) && 0 !== ($curlopts[\CURLOPT_INFILESIZE] ?? null)) {
                     $curlopts[\CURLOPT_HTTPHEADER][] = 'Content-Type: application/x-www-form-urlencoded';
                 }
             }

